accord-core/src/main/java/accord/impl/InMemoryCommandStore.java

```diff
@@ -388,20 +395,32 @@ private void loadCommandsForKey(RoutableKey key,
         timestampsForKey.put(key, timestampsForKey((Key) key).createSafeReference());
     }
 
+    protected void validateRead(Command current) {}
+
     protected final InMemorySafeStore createSafeStore(PreLoadContext context, RangesForEpoch ranges)
     {
         Map<TxnId, InMemorySafeCommand> commands = new HashMap<>();
         Map<RoutableKey, InMemorySafeCommandsForKey> commandsForKey = new HashMap<>();
         Map<RoutableKey, InMemorySafeTimestampsForKey> timestampsForKey = new HashMap<>();
 
         context.forEachId(txnId -> commands.put(txnId, lazyReference(txnId)));
+        for (InMemorySafeCommand safe : commands.values())
+        {
+            GlobalCommand global = safe.unsafeGlobal();
+            if (global == null) continue;
+            Command current = global.value();
+            if (current == null) continue;
+            validateRead(current);
+        }
 
         for (Seekable seekable : context.keys())
         {
             switch (seekable.domain())
             {
                 case Key:
                     RoutableKey key = (RoutableKey) seekable;
+                    if (context.keyHistory() == KeyHistory.NONE)
```

Comment:
Currently in C* we only load _either_ timestamps _or_ commands. So we should apply the same logic here too.

Reply:
switched to

```
switch (context.keyHistory())
                    {
                        case NONE:
                            continue;
                        case COMMANDS:
                            commandsForKey.put(key, commandsForKey((Key) key).createSafeReference());
                            break;
                        case TIMESTAMPS:
                            timestampsForKey.put(key, timestampsForKey((Key) key).createSafeReference());
                            break;
                        default: throw new UnsupportedOperationException("Unknown key history: " + context.keyHistory());
                    }
```